File: Sample/Sample/Track.m
//
//  Track.m
//  Sample
//
//  Created by DebugHolic on 31/03/2019.
//  Copyright © 2019 Sidekick-Academy. All rights reserved.
//

#import "Track.h"

@implementation Track

+ (Track *)trackWithMetadata:(SAAudioMetadata *)metadata {
    return [self.class trackWithMetadata:metadata albumArt:nil];
}

+ (Track *)trackWithMetadata:(SAAudioMetadata *)metadata albumArt:(UIImage *)image {
    Track *track = [[Track alloc] init];
    if (metadata.title) {
        track.title = [NSString stringWithString:metadata.title];
    }
    if (metadata.artist) {
        track.artist = [NSString stringWithString:metadata.artist];
    }
    if (metadata.album) {
        track.album = [NSString stringWithString:metadata.album];
    }
    track.duration = metadata.duration;
    track.bitdepth = metadata.bitdepth;
    track.samplerate = metadata.samplerate;
    track.albumArt = image;
    return track;
}

- (void)setDuration:(NSUInteger)duration {
    _duration = duration;
}

- (void)setAlbum:(NSString * _Nullable)album {
    _album = album;
}

- (void)setTitle:(NSString * _Nullable)title {
    _title = title;
}

- (void)setArtist:(NSString * _Nullable)artist {
    _artist = artist;
}

- (void)setBitdepth:(NSUInteger)bitdepth {
    _bitdepth = bitdepth;
}

- (void)setSamplerate:(NSUInteger)samplerate {
    _samplerate = samplerate;
}

- (void)setAlbumArt:(UIImage * _Nullable)albumArt {
    _albumArt = albumArt;
}

@end

File: Sample/Sample/Track.swift
//
//  Track.swift
//  Sample
//
//  Created by 김영훈 on 12/6/24.
//  Copyright © 2024 Sidekick-Academy. All rights reserved.
//

import Foundation

struct Track: AudioPlayable {
    let mediaInfo: MediaInfo?
    let albumArt: UIImage?
    let path: String
    
    init(_ path: String) {
        self.path = path
        self.mediaInfo = MediaInfo(metadata: MetadataExtractor.metadata(withPath: path))
        self.albumArt = AlbumArtExtractor.albumArt(withPath: path)
    }
}

struct MediaInfo {
    let title: String?
    let album: String?
    let artist: String?
    let creator: String?
    let date: String?
    let samplerate: UInt
    let bitdepth: UInt
    let channels: UInt
    let duration: UInt

    init(title: String? = nil, album: String? = nil, artist: String? = nil, creator: String? = nil, date: String? = nil, samplerate: UInt? = nil, bitdepth: UInt? = nil, channels: UInt? = nil, duration: UInt?) {
        self.title = title
        self.album = album
        self.artist = artist
        self.creator = creator
        self.date = date
        self.samplerate = samplerate ?? 0
        self.bitdepth = bitdepth ?? 0
        self.channels = channels ?? 0
        self.duration = duration ?? 0
    }
    
    init(metadata: AudioMetadata?) {
        let title = metadata?.title
        let album = metadata?.album
        let artist = metadata?.artist
        let creator = metadata?.creator
        let date = metadata?.date
        let samplerate = metadata?.samplerate
        let bitdepth = metadata?.bitdepth
        let channels = metadata?.channels
        let duration = metadata?.duration
        
        self.init(title: title, album: album, artist: artist, creator: creator, date: date, samplerate: samplerate, bitdepth: bitdepth, channels: channels, duration: duration)
    }
}
